### PR TITLE
support starcoder2

### PR DESCRIPTION
**Pull request by grimoire**
Support starcoder2 https://huggingface.co/bigcode/starcoder2-7b

### CODE REVIEW

**Review by AllentDan — COMMENTED** [review]
What is the expected version of transformers to run the model? I got 
```
ERROR - ValueError: The checkpoint you are trying to load has model type `starcoder2` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.
2024-04-24 12:48:26,763 - lmdeploy - ERROR - <transformers> test failed!
Load model config with transformers==4.38.2 failed. Please make sure model can be loaded with transformers API.
```

**Comment by grimoire** [comment]
4.40  I guess.

**Review by AllentDan — COMMENTED** [review]
Got
```
double enter to end input >>> hi

hi _1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000000000", null ],
    [ "GetID", "class_hi_1_1_i_d_manager.html#a00000000000000000000000000
```
for the command `lmdeploy chat torch /path/of/starcoder2/startcoder2-7b`.

transformers version 4.40.1

**Review by AllentDan — COMMENTED** [review]
May resolve the conflicts

**Comment by grimoire** [comment]
@AllentDan 
This is not a chat model, this output might be expected.

official demo:
```
inputs = tokenizer.encode("def print_hello_world():", return_tensors="pt").to(device)
outputs = model.generate(inputs)
```

**Review by AllentDan — APPROVED** [review]
LGTM

File: lmdeploy/model.py [hunk 1]
```diff
@@ -1299,6 +1299,23 @@ def match(cls, model_path: str) -> Optional[str]:
             return 'internvl-zh-hermes2'
 
 
+@MODELS.register_module(name='starcoder2')
```
Comment (lvhan028) [inline lmdeploy/model.py:1302]: If it is not a chat model or an instruct model, there is no need to add a template. 
`best_match_model` will fallback to base template

**Pull request merged**